src/Security/TokenAuthenticator.php

```diff
@@ -0,0 +1,191 @@
+    private $clientId;
+    private $clientSecret;
+    private $endPoint;
+
+    /**
+     * TokenAuthenticator constructor.
+     *
+     * @param string $bindOpenplatformId
+     * @param string $bindOpenplatformSecret
+     * @param string $bindOpenplatformIntrospectionUrl
+     * @param AdapterInterface $tokenCache
+     * @param HttpClientInterface $httpClient
+     */
+    public function __construct(string $bindOpenplatformId, string $bindOpenplatformSecret, string $bindOpenplatformIntrospectionUrl, AdapterInterface $tokenCache, HttpClientInterface $httpClient)
+    {
+        $this->client = $httpClient;
+        $this->cache = $tokenCache;
+
+        $this->clientId = $bindOpenplatformId;
+        $this->clientSecret = $bindOpenplatformSecret;
+        $this->endPoint = $bindOpenplatformIntrospectionUrl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        return $request->headers->has('authorization');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCredentials(Request $request)
+    {
+        return $request->headers->get('authorization');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUser($credentials, UserProviderInterface $userProvider)
+    {
+        if (null === $credentials) {
+            // The token header was empty, authentication fails with HTTP Status
+            // Code 401 "Unauthorized"
+            return null;
+        }
+
+        // Parse token information from the bearer authorization header.
+        preg_match('/Bearer\s(\w+)/', $credentials, $matches);
+        if (2 !== count($matches)) {
+            return null;
+        }
+
+        $token = $matches[1];
+
+        // Try getting item from cache.
+        $item = $this->cache->getItem($token);
+        if ($item->isHit()) {
+            return $item->get();
+        }
+
+        try {
+            $response = $this->client->request(
+                'POST',
+                $this->endPoint.'?access_token='.$token,
+                [
+                    'auth_basic' => [$this->clientId, $this->clientSecret],
+                ]
+            );
+
+            if (200 !== $response->getStatusCode()) {
+                return null;
+            }
+
+            $content = $response->getContent();
+            $data = json_decode($content);
+
+            // Token not valid, hence not active at the introspection end-point.
+            if (false === $data->active) {
+                return null;
+            }
+        } catch (HttpExceptionInterface $e) {
+            return null;
+        } catch (ExceptionInterface $e) {
+            return null;
+        }
+
+        // Create user object.
+        $tokenExpireDataTime = new \DateTime($data->expires, new \DateTimeZone('Europe/Copenhagen'));
+        $user = new User();
+        $user->setPassword($token);
+        $user->setExpires($tokenExpireDataTime);
+        $user->setAgency($data->agency);
```

Comment:
If this was not an agency token but an user token we need to detected that as I don't think an user token has agency on it. So this `$data->agancy` would result in a 500 error.

We may need to look at detecting token type an give an invalided token or access denied (return null)

Reply:
Fixed